crmsh/ui_cluster.py

```diff
@@ -91,78 +138,94 @@ def __init__(self):
-                return
-            bootstrap.start_pacemaker()
-            if utils.is_qdevice_configured():
-                utils.start_service("corosync-qdevice")
-            logger.info("Cluster services started")
-        except IOError as err:
-            context.fatal_error(str(err))
-
-        # TODO: optionally start services on all nodes or specific node
+        node_list = parse_option_for_nodes(context, *args)
+        for node in node_list[:]:
+            if utils.service_is_active("pacemaker.service", remote_addr=node):
+                logger.info("Cluster services already started on {}".format(node))
+                node_list.remove(node)
+        if not node_list:
+            return
+
+        bootstrap.start_pacemaker(node_list)
+        if utils.is_qdevice_configured():
+            utils.start_service("corosync-qdevice", node_list=node_list)
+        for node in node_list:
+            logger.info("Cluster services started on {}".format(node))
 
     @command.skill_level('administrator')
-    def do_stop(self, context):
+    def do_stop(self, context, *args):
         '''
-        Stops the cluster services on this node
+        Stops the cluster services on all nodes or specific node(s)
         '''
-        try:
-            if not utils.service_is_active("corosync.service"):
-                logger.info("Cluster services already stopped")
-                return
-            if utils.service_is_active("corosync-qdevice"):
-                utils.stop_service("corosync-qdevice")
-            utils.stop_service("corosync")
-            logger.info("Cluster services stopped")
-        except IOError as err:
-            context.fatal_error(str(err))
-
-        # TODO: optionally stop services on all nodes or specific node
+        node_list = parse_option_for_nodes(context, *args)
+        for node in node_list[:]:
+            if not utils.service_is_active("corosync.service", remote_addr=node):
+                if utils.service_is_active("sbd.service", remote_addr=node):
+                    utils.stop_service("corosync", remote_addr=node)
+                    logger.info("Cluster services stopped on {}".format(node))
+                else:
+                    logger.info("Cluster services already stopped on {}".format(node))
+                node_list.remove(node)
+        if not node_list:
+            return
+
+        bootstrap.wait_for_cluster("Waiting for {} online".format(' '.join(node_list)), node_list)
+
+        # When dlm configured and quorum is lost, before stop cluster service, should set
+        # enable_quorum_fencing=0, enable_quorum_lockspace=0 for dlm config option
+        if utils.is_dlm_configured(node_list[0]) and not utils.is_quorate(node_list[0]):
+            logger.debug("Quorum is lost; Set enable_quorum_fencing=0 and enable_quorum_lockspace=0 for dlm")
+            utils.set_dlm_option(peer=node_list[0], enable_quorum_fencing=0, enable_quorum_lockspace=0)
```

Comment:
IIUC, both `enable_quorum_fencing` and `enable_quorum_lockspace` need to be disabled for an inquorate dlm to gracefully stop? Could there be any risks?

Reply:
yes, both. 
from my opinion, there doesn't have any risk. because this action is only triggered by cmd "crm cluster stop".
before this patch, cluster (by dlm_controld) deny to do any further action until quorum is true.
after this patch, cluster will directly stop, there is no other behavior.

Reply:
The biggest concern is, given that this cluster partition is inquorate, there might a quorate partition standing... For example this partition has been split and inquorate, but somehow hasn't been fenced. If we are shutting down this cluster partition, and in case these settings make the inquorate partition be able to acquire the access to lockspace and corrupt data, that would be a disaster. We should never sacrifice data integrity even for graceful shutdown...

Reply:
enable_quorum_lockspace is disabled, this will make dlm lockspace related operation can keep going when the cluster quorum is lost.

Reply:
your comment makes sense to me.

Reply:
your comment makes sense to me.
the better solution should add a prompt info for user, let user to make decision: whether crm modify enable_quorum_(fencing|lockspace) to graceful shutdown.

Reply:
the better solution should add a prompt info for user, let user to make decision: whether crm modify enable_quorum_(fencing|lockspace) to graceful shutdown.

Reply:
> the better solution should add a prompt info for user, let user to make decision: whether crm modify enable_quorum_(fencing|lockspace) to graceful shutdown.

It's good idea. We should give clear heads-up/warning so that they know what they are doing.

Reply:
> The biggest concern is, given that this cluster partition is inquorate, there might a quorate partition standing... For example this partition has been split and inquorate, but somehow hasn't been fenced. If we are shutting down this cluster partition, and in case these settings make the inquorate partition be able to acquire the access to lockspace and corrupt data, that would be a disaster. We should never sacrifice data integrity even for graceful shutdown...

No risk here. These changes, "enable_quorum_fencing=0 and enable_quorum_lockspace=0", are intentionally put as the runtime by "crm cluster stop". It will revert back to the static default config after restart the dlm_controld daemon.

With that, it looks to me an INFO is enough to tell those runtime changes happens, and probably not as WARNING even.

Reply:
We don't know whether one *really* is aware of the real situation about whether there's no other cluster partitions standing...

Reply:
Thinking about it on the more safe/paranoid side, even if the user has confirmed to proceed, but allowing this simultaneously on multiple nodes is more like opening a Pandora's box... Since right after that,  during stop, this cluster partition might spit apart into more partitions...

If we go for it, we could ask user once, but we'd better proceed this specific procedure the serialized way: so
`set_dlm_option -> stop dlm/pacemaker` for only one node at a time, and proceed for another node only after it has succeeded on this one.

Reply:
My previous point stays with the last standing single node situation. Given that the node is inquorate already, the proposed behavior for `crm cluster stop` (aka. crmsh -> set_dlm_option -> stop dlm/patcemaker) is just a "fencing" operation to protect data integrity, though no necessarily STONITH at the node level to do reboot. 

Well, for the situation with multiple inquorate nodes, aka. multiple partitions, then the code here do have problem for '--all' situation. Simply because set_dlm_option only applies to one node. Not sure, if it is simple enough to address it in this PR, or open an issue to clarify this in another PR.

Reply:
set_dlm_option is implemented by "dlm_tool set_config", which only run on a single node each time.

Reply:
> set_dlm_option is implemented by "dlm_tool set_config", which only run on a single node each time.

Fine. 

Given the above situation, when all nodes are inquorate, what's the suggested graceful shutdown steps internally for "stop --all" ? My reading of the code is it only change the current local node, no repeat the same on other nodes. 

The situation is more fun, in theory for a big cluster, some nodes are quorate, some are not. What's the suggested internal steps for "stop --all"? @zhaohem

Reply:
Maybe crmsh should never operate a cluster in the transient state at all by default? And ask user to answer Y/n, or user --force ?